File: python/groupAnagram.py
# Time -> 0(N*N*M) Space -> 0(N*W) where N = number of words and M = longest word
def groupAnagram(words: list) -> list:
    map = {}
    wordMap = {}
    wordsMap = {}
    initStore = []
    result = []
    
    for word in words:
        wordsMap[word] = False
    
    for word in words:
        map = {}
        initStore = []
        if wordsMap[word] == True:
            continue
        map = getMap(map, word)
        initStore.append(word)
        
        for newWord in words:
            if wordsMap[newWord] == True or newWord == word or len(newWord) != len(word):
                continue
            wordMap = {}
            wordMap = getMap(wordMap, newWord)
            if checkAnagram(map, wordMap):
                initStore.append(newWord)
                wordsMap[newWord] = True
                
        result.append(initStore)
        wordsMap[word] = True
    return result


def getMap(map: dict, word: str) -> dict:
    for char in word:
        map[char] = map.get(char, 0) + 1
    return map
        
        
def checkAnagram(firstMap: dict, secondMap: dict) -> bool:
    if len(firstMap) != len(secondMap):
        return False
    
    for k, v in firstMap.items():
        if k not in secondMap:
            return False
        val = secondMap[k]
        if val != v:
            return False
        
    return True


# Time -> 0(N*M log(M)) Space -> 0(MN)
def groupAnagram1(words: list) -> list:
    map = {}
    result = []
    for i in range(len(words)):
        word = sorted(words[i])
        word = ''.join(word)
        if word in map:
            map[word].append(words[i])
        else:
            map[word] = [words[i]]
    

    return list(map.values())
        
print(groupAnagram(["yo", "bro", "oy", "wrong", "cat","tac", "act"]))
print(groupAnagram1(["yo", "bro", "oy", "wrong", "cat", "tac", "act"]))

File: go/groupAnagram/groupAnagram.go
package main

import (
	"fmt"
	"sort"
)

// Time -> 0(N*N*M) Space -> 0(N*W) where N = number of words and M = longest word
func groupAnagram(words []string) [][]string {
	set := make(map[string]int)
	wordSet := make(map[string]int)
	wordsSet := make(map[string]bool)
	initStore := make([]string, 0)
	result := make([][]string, 0)


	for _, val := range words {wordsSet[val] = false}

	for _, word := range words {
		if seen, _ := wordsSet[word]; seen {continue}
		initStore = make([]string, 0)
		set = make(map[string]int)
		set = getMaps(set, word)
		initStore = append(initStore, word)
		for _, newWord := range words {
			wordSet = make(map[string]int)
			if seen, _ := wordsSet[word]; seen || newWord == word{continue}
			if len(word) != len(newWord) {continue}
			wordSet = getMaps(wordSet, newWord)
			if checkAnagram(set, wordSet) {
				initStore = append(initStore, newWord)
				wordsSet[newWord] = true
			}
		}
		result = append(result, initStore)
		wordsSet[word] = true
	}
	return result
}

func getMaps(set map[string]int, word string) map[string]int {
	for _, v := range word {
		set[string(v)]++
	}
	return set
}

func checkAnagram(firstSet, secondSet map[string]int) bool{
    if len(firstSet) != len(secondSet) {return false}
	for k, v := range firstSet {
		if val, exists := secondSet[k]; !exists || val != v {return false}
	}
	return true
}


// Time -> 0(N*M log(M)) Space -> 0(MN)
func groupAnagram1(words []string) [][]string {
	set := make(map[string][]string)
	result := make([][]string, 0)

	for i := range words {
		word := sortWords(words[i])
		if val, exists := set[word]; exists {
			val = append(val, words[i])
			set[word] = val
		}else {set[word] = []string{words[i]}}
	}

	for _, v := range set {result = append(result, v)}
	return result
}


func sortWords(word string) string {
	wordbyte := []byte(word)
	sort.Slice(wordbyte, func(i, j int) bool {
		return wordbyte[i] < wordbyte[j]
	})
	return string(wordbyte)
}


func main() {
	fmt.Println(groupAnagram([]string{"yo", "bro", "oy", "wrong", "cat","tac", "act"}))
	fmt.Println(groupAnagram1([]string{"yo", "bro", "oy", "wrong", "cat","tac", "act"}))
}
